source/scripts/data/monsters.js

```diff
@@ -31,4 +31,206 @@ export default {
+                    delete choices[choices.indexOf({x: +1})]
+            }
+            for(var monster of this.game.monsters) {
+                if (monster.position.y == this.position.y - 1 && monster.position.x == this.position.x)
+                    delete choices[choices.indexOf({y: -1})]
+            }
+            for(var monster of this.game.monsters) {
+                if (monster.position.y == this.position.y + 1 && monster.position.x == this.position.x)
+                    delete choices[choices.indexOf({y: +1})]
+            }
+
+            if (this.position.x == 0)
+                delete choices[choices.indexOf({x: -1})]
+            if (this.position.x == DATA.FRAME.WIDTH)
+                delete choices[choices.indexOf({x: +1})]
+            if (this.position.y == 0)
+                delete choices[choices.indexOf({y: -1})]
+            if (this.position.y == DATA.FRAME.HEIGHT)
+                delete choices[choices.indexOf({y: +1})]
+
+            return choices[Math.floor((Math.random() * choices.length))]
+        }
+    },
+    BLUE_BAT: {
+        sprite: DATA.SPRITES.MONSTERS.BAT,
+        color: DATA.COLORS.BLUE,
+        health: 1,
+        strength: 1,
+        movement: function () {
+            if (this.position.x < 0) return {x: +1}
+            if (this.position.x >= DATA.FRAME.WIDTH) return {x: -1}
+            if (this.position.y < 0) return {y: +1}
+            if (this.position.y >= DATA.FRAME.HEIGHT) return {y: -1}
+
+            var choices = [
+                {x: -1},
+                {x: +1},
+                {y: -1},
+                {y: +1}
+            ]
+
+            for(var monster of this.game.monsters) {
+                if (monster.position.x == this.position.x - 1 && monster.position.y == this.position.y)
+                    delete choices[choices.indexOf({x: -1})]
+            }
+            for(var monster of this.game.monsters) {
+                if (monster.position.x == this.position.x + 1 && monster.position.y == this.position.y)
+                    delete choices[choices.indexOf({x: +1})]
+            }
+            for(var monster of this.game.monsters) {
+                if (monster.position.y == this.position.y - 1 && monster.position.x == this.position.x)
+                    delete choices[choices.indexOf({y: -1})]
+            }
+            for(var monster of this.game.monsters) {
+                if (monster.position.y == this.position.y + 1 && monster.position.x == this.position.x)
+                    delete choices[choices.indexOf({y: +1})]
+            }
+
+            if (this.position.x == 0)
+                delete choices[choices.indexOf({x: -1})]
+            if (this.position.x == DATA.FRAME.WIDTH)
+                delete choices[choices.indexOf({x: +1})]
+            if (this.position.y == 0)
+                delete choices[choices.indexOf({y: -1})]
+            if (this.position.y == DATA.FRAME.HEIGHT)
+                delete choices[choices.indexOf({y: +1})]
```

Comment:
Until then, let's keep it the way you've got; an array of possible movements which is then pruned if a) it would collide with a wall, or b) it would collide with a monster. Just refactor this into a function that accepts an array of possible movements (or defaults to orthogonal neighbors), and returns that same array but pruned for collisions. (Just so this code isn't repeated all over the place.)